File: 2020/src/day01.cpp
#include "day01.h"
#include <ranges>

ResultBase* Day01::SolveFirst() const
{
	std::vector<int> numbers = ParseInput();
	for(int i = 0; i < numbers.size(); i++)
	{
		for(int k = i + 1; k < numbers.size(); k++)
		{
			if(numbers[i] + numbers[k] == 2020)
			{
				return new Result(numbers[i] * numbers[k]);
			}
		}
	}
	return nullptr;
}

ResultBase* Day01::SolveSecond() const
{
	std::vector<int> numbers = ParseInput();
	for(int i = 0; i < numbers.size(); i++)
	{
		for(int k = i + 1; k < numbers.size(); k++)
		{
			for(int l = k + 1; l < numbers.size(); l++)
			{
				if(numbers[i] + numbers[k] + numbers[l] == 2020)
				{
					return new Result(numbers[i] * numbers[k] * numbers[l]);
				}
			}
		}
	}
	return nullptr;
}

std::vector<int> Day01::ParseInput() const
{
	auto res = input.lines | std::views::transform([](const std::string& str)
	{
		return std::stoi(str);
	});
	return std::vector<int>(res.begin(), res.end());
}

File: 2025/src/day01.fs
module Day01
open Input
open Common

[<Literal>]
let name = "Secret Entrance"

let (%!) a b = if a < 0 then (a % b + b) % b else a % b

let parseInput (input: Input) = 
    input.lines
    |> List.map (fun n -> (int n.[1..]) * (if n.[0] = 'L' then -1 else 1))

let solveFirst (input: Input) = 
    input
    |> parseInput
    |> List.scan (fun acc n -> (acc + n) %! 100) 50
    |> List.filter ((=) 0)
    |> List.length

let solveSecond (input: Input) = 
    let rec move curr count = function
        | head :: tail -> 
            let next = curr + head
            let crosses = abs (next / 100) + (if (sign next) <> (sign curr) && (sign curr) <> 0 then 1 else 0)
            move (next %! 100) (count + crosses) tail
        | [] -> count

    move 50 0 (parseInput input)
